계산기 I [STEP 2] 마이노

Mino777 opened the pull request:
안녕하세요 스티븐! @stevenkim18 
STEP2 PR입니다!
잘 부탁드립니다🥰

---

### UML
![https://s3.ap-northeast-2.amazonaws.com/media.yagom-academy.kr/resources/6131c8fa2e11413823f8dd7f/6189d7537c82755a83c68a7d.jpg](https://s3.ap-northeast-2.amazonaws.com/media.yagom-academy.kr/resources/6131c8fa2e11413823f8dd7f/6189d7537c82755a83c68a7d.jpg)

### 고민한 점 및 궁금한 점
1.  UML을 토대로 요구사항 파악
- UML만 보고 기능을 명세해야되는 과정을 처음해보다보니 이 부분에서 시간을 가장 많이 쓴 것 같습니다😭
-   `Operator`: 연산자 정의, 매칭 및 실질적인 연산을 담당한다고 생각했습니다.
-   `ExpressionParser`: 표현식을 만드는것과 관련된 메서드를 모아둔 타입이라고 생각했습니다.
    -   `parse`: 계산기에서 사용자가 입력했던 요소들을 해당 메서드를 호출하는 부분에서 저장해두었다가, =을 눌렀을 때 호출하여 input을 받아서 operand, operator를 분리한 후 각 queue에 enqueue해주고 이를 Formular에 주입해서 반환해주는 메서드라고 생각했습니다.
    -   `componentsByOperators`: operand와 operator를 분리해주는 메서드라고 생각했습니다.
-   `Formula`
    -   `result`: 계산기에서 =을 눌렀을 때 큐에 입력된 연산들을 수행하는 메서드라고 생각했습니다.
- 스티븐과 협의 후 추가로 진행한 메서드
- `ExpressionParser`
	- `componentsByOperands`: `componentsByOperators`의 경우 Operand만 뽑아내는 메서드이고, 해당 메서드의 경우 Operator만 뽑아내는 메서드입니다.
	- `makeOperandsQueue`: queue에 enqueue해주는 과정을 분리해주어 반환값으로 `CalculatorItemQueue<Double>`을 반환해줍니다.
	- `makeOperatorsQueue`: queue에 enqueue해주는 과정을 분리해주어 반환값으로 `CalculatorItemQueue<Operator>`를 반환해줍니다.
- `componentsByOperators`의 경우 반환값을 Double로 진행하면 한뎁스를 줄일 수 있지만, 우선 UML을 바꾸지 않는 선에서 진행해보았습니다.

2. TDD
- TDD를 위해 의존성 관계 그래프의 최말단 타입부터 구현 해보았습니다!
	- Operator -> Formula -> ExpressionParser 순
- 이번 스텝에서 TDD를 하면서 왜 개발시간이 훨씬 더 늘어나는지 아주 확실하게 알게 된 것 같습니다😭 그래도 확실히 안정적으로 자신감있게 한줄 한줄 개발할 수 있게 해주는 것은 틀림없는 것 같습니다!
- 요구사항 중 0으로 나누는 경우 에러처리를 진행해줘야 하는 부분이 있었는데, 해당 부분의 경우 Operator에서 부터 계속 진행이 되다보니 모든 단계에서 테스트를 진행하게 되었습니다.
- 이런식으로 중첩되는 테스트의 경우 모든 단계에서 하는것이 맞을까요? 아니면 최상위 메서드에서만 테스트하는게 맞을까요?
- 그런데 생각해보니 TDD의 경우에 당연히 해당 테스트를 진행해야 다음 단계로 갈 수 있는 것 같다는 생각이 들어서 모든 단계에서 해주는게 맞는 것 같습니당 스티븐의 생각은 어떠실까요?!

3. 메서드 분리
- 말씀드렸던 `ExpressionParser` 타입에서 최대한 메서드를 분리해보려고 진행을 해보았는데, 뭔가 분리하면 할수록 복잡성이 올라가는 느낌이 드는 것 같습니다. 기분 탓일까요? 그런데 분리를 안하면 뭔가 하나의 메서드가 역할이 많아지는 것 같고 그렇습니다😭 조언 부탁드립니당!

감사합니다!!🙇‍♂️

stevenkim18 commented on Calculator/Calculator/Extension/String.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+extension String {
+    func split(with target: Character) -> [String] {
+        return self.split(separator: target)
+            .map { character in
+                String(character)
+            }
+    }
+}
```
만약 계속 고차함수를 사용하실것이라고 하면 
```swift
.map {String($0)}
.filter {...}
```
이런 식으로 한줄씩 쓰면 가독성이 올라갈 것 같아요!! 어떻게 생각하시나요?

Mino777 replied:
캠프 초반 프로젝트에서 처음에 익숙해질때까지 고차함수를 축약하지 않고 사용해보라고 피드백을 받은적이 있어서 무의식적으로 풀어서 사용한 것 같습니다. 앞으론 축약해서 사용하겠습니다! 

감사합니다!

stevenkim18 replied:
저는 그렇게 쓴다는 것이지 꼭 그렇게 쓰라는 것은 아니에요 ㅎㅎ 축약형이 좋을 때가 있고 아닐떄가 있습니다!! 때에 따라 알맞게 사용하셔야 돼요!!

stevenkim18 commented on Calculator/Calculator/Model/ExpressionParser.swift:
```diff
@@ -0,0 +1,68 @@
+    
+    private static func componentsByOperators(from input: String) -> [String] {
+        input.split(with: " ")
+            .compactMap { string in
+                Double(string)
+            }
+            .map { double in
+                String(double)
+            }
+    }
+    
+    private static func componentsByOperands(from input: String) -> [Operator] {
+        let operatorTypes = Operator
+            .allCases
+            .map { type in
+                String(type.rawValue)
+            }
+        
+        return input.split(with: " ")
+            .filter { string in
+                operatorTypes.contains(string)
+            }
+            .compactMap { string in
+                Operator(rawValue: Character(string))
+            }
+    }
+    
+    private static func makeOperandsQueue(from input: String) -> CalculatorItemQueue<Double> {
+        var operands = CalculatorItemQueue<Double>()
+        
+        componentsByOperators(from: input)
+            .compactMap { string in
+                Double(string)
+            }
+            .forEach { operand in
+                operands.enqueue(operand)
+            }
+        
+        return operands
+    }
```
- 위에도 말씀 드렸지만 고차함수를 많이 사용하셔서 오히려 가독성이 떨어지는 것 같아요!
- 고차함수를 사용하고 안하고의 장단점은 무엇이 있을까요?
- map filter compactMap이 무엇인지 각각 설명해주세요!

Mino777 replied:
#### 고차함수 장점

- 좀 더 적은 수의 코드로 동일한 결과를 얻을 수 있습니다.
- 이번 프로젝트에 쓴거처럼 남용하지 않는다면 가독성도 높일 수 있습니다.
- for문 처럼 구문 안에서 변수를 변경하는 작업이 필요없이 완성된 결과를 리턴해주기 때문에 사이드 이펙트도 줄일 수 있을 것 같습니다.
- 위와 같은 이유 때문에 함수형 프로그래밍을 진행할 때 적극 활용될 수 있을 것 같습니다.

#### 고차함수 단점

- 남용하는 경우 지금처럼 가독성이 오히려 떨어질 수 있을 것 같습니다.
- 체이닝이 깊어지는 경우에 컴파일러단에서 최적화가 이루어진다 해도, 만약 for문 한개로 진행할 수 있는 작업의 경우 성능적으로 밀릴 것 같습니다.

#### map
map의 경우 기존의 Sequence 타입의 데이터를 변형하여 새로운 컨테이너를 생성해 반환해줍니다. 

#### filter
filter의 경우 컨테이너 내부에서 조건에 따라 값을 걸러서 새로운 컨테이너로 추출해줍니다.

#### compactMap
compactMap의 경우 map과 동일한 작업을 수행하지만, nil이 아닌 요소만 반환해줍니다.

고차함수들을 직접 구현해보았습니다.
```swift
extension Sequence {
    func myMap<T>(conversion: (Element) -> T ) -> [T] {
        var result: [T] = []
        
        self.forEach({ element in
            result.append(conversion(element))
        })
        return result
    }
    
    func myFilter(selected: (Element) -> Bool) -> [Element] {
        var result: [Element] = []
        
        self.forEach({ element in
            if selected(element) {
                result.append(element)
            }
        })
        return result
    }
    
    func myCompactMap<T>(conversion: (Element)-> T?) -> [T] {
        var result = [T]()
        forEach { element in
            if let convertedEntity = conversion(element) {
                result.append(convertedEntity)
            }
        }
        return result
    }
    
    func myReduce<T>(initValue: T, calculation: (T, Element) -> T) -> T {
        var result: T = initValue
        self.forEach({
            result = calculation(result, $0)
        })
        return result
    }
}
```

stevenkim18 replied:
와우 직접 구현까지 ㅎㅎㅎ

stevenkim18 commented on Calculator/Calculator/Model/Formula.swift:
```diff
@@ -0,0 +1,49 @@
+        return operators.count
+    }
+    
+    init(operands: CalculatorItemQueue<Double>,
+         operators: CalculatorItemQueue<Operator>) {
+        self.operands = operands
+        self.operators = operators
+    }
```
구조체에서 프로퍼티들을 private으로 해도 자동으로 이니셜라이져가 생성되지 않나요? 
저도 헷갈려서 여쭤봅니다!

Mino777 replied:
'Formula' initializer is inaccessible due to 'private' protection level 이라는 에러가 나옵니다!

멤버 변수가 private일시에 initializer도 접근제어레벨이 private이 되어 멤버와이즈 이니셜라이저를 사용하지 못한다고 알고 있습니다!

감사합니다!

stevenkim18 replied:
아 맞아요 private은 맴버 이니셜라이져가 안되는군요 ㅎㅎ

stevenkim18 commented on Calculator/CalculatorItemQueueTests/ExpressionParserTests.swift:
```diff
@@ -0,0 +1,58 @@
+import XCTest
+
+class ExpressionParserTests: XCTestCase {
+    var mockExpression = "1 + 2 - 3 / 4 * -5"
```
- mock은 어떤 의미인가요?, 언제 주로 사용하나요? 
- 더 다양한 수식을 테스트 해보셨으면 좋았을 것 같은데 이렇게 하나만 프로퍼티로 뺴주셔서 테스트 하신 이유가 있을까요?

Mino777 replied:
mock같은 경우 주로 실제 객체와 비슷한 가짜 객체를 만들어서 테스트를 진행할 때 사용한다고 알고있습니다. 여기서는 그냥 가짜라는 의미를 붙여 사용을 했었는데, 이런 프로퍼티의 경우 그냥 expression 이런식으로 기본적인 네이밍을 해서 사용하는게 좋을까요?

해당 테스트에서는 사칙연산의 순서를 무시한 결과값만 테스트하면 나머지 수식에 관한 테스트 또한 커버가 가능하다고 생각하였는데, 이런 생각은 굉장히 위험한 것 같습니다. 당연히 다양한 수식으로 테스트를 진행해야할 것 같습니다. 앞으로 주의하겠습니다!

감사합니다!

stevenkim18 replied:
지금은 mock이 한개여서 괜찮지만 저는 많아지면 따로 관리해주었습니다.

엣지케이스가 에러케이스가 저 있으면 좋을 것 같아요 
- 연산자는 없고 숫자만 오는 경우
- 엄청 큰 숫자를 넣을 때(?)

등등 생각해볼 수 있는 케이스는 많을 것 같아요 ㅎㅎ

stevenkim18 commented on Calculator/CalculatorItemQueueTests/FormulaTests.swift:
```diff
@@ -0,0 +1,94 @@
+
+    func test_result를호출할때_결과값과예상값이같아야한다() {
+        // given
+        var mockOperandQueue = CalculatorItemQueue<Double>()
+        var mockOperatorQueue = CalculatorItemQueue<Operator>()
```
queue들이 계속 사용되고 있네요!!
프로퍼티로 뺴주면 어떨까요?

Mino777 replied:
앗 바보같이 계속 사용하고 있었네요..!! 빼주었습니다! 
감사합니다!

stevenkim18 replied:
보통 2개이상 메소드에서 사용되면 저는 빼주는 편이에요 ㅎㅎ

stevenkim18 commented on Calculator/CalculatorItemQueueTests/OperatorTests.swift:
```diff
@@ -0,0 +1,82 @@
+        sut = nil
+    }
+    
+    func test_Operator가add이고calculate를호출할때_결과값이예상값과같아야한다() {
```
제 개인적인 생각은 테스트 이름 규칙을 모두 똑같이 적용할 필요는 없다고 생각합니다!!
그냥 여기서는 `add`를 테스트 한다고만 알려줘도 좋을 것 같아요 ㅎㅎ `결과값이예상값과같아야한다`가 계속 반복되고 있네요 ㅎㅎ

stevenkim18 replied:
Xcode 왼쪽에 테스트 탭이 있는데 
거기에는 테스트 이름만 나와요 
간단히 테스트 이름만을 보고도 명세가 될 수 있을 것 같아서 
저도 떄에 따라서 조금 더 명확하고 간결하게 쓰려고 하고 있어요 ㅎㅎ

stevenkim18 commented on Calculator/CalculatorItemQueueTests/OperatorTests.swift:
```diff
@@ -0,0 +1,82 @@
+        XCTAssertEqual(result, expected)
+    }
+    
+    func test_Operator가substract이고calculate를호출할때_결과값이예상값과같아야한다() {
+        // given
+        sut = .substract
+        
+        // when
+        let result = try? sut.calculate(lhs: 2.0, rhs: 1.0)
+        
+        // then
+        let expected = 1.0
+        XCTAssertEqual(result, expected)
+    }
```
```swift
// given
let lhs = 2
let rhs = 1

...
//then
XCTAssertEqual(result, lhs-rhs)
```

위와 같이 operator를 거친것과 직접 계산한 값을 테스트 해보는 것이 더 의미 있는 것 같아요
`expected` 라는 기대값은 주기 나름이기 떄문에 위에 같이 하는 건 어떻게 생각하시나요?

Mino777 replied:
이렇게 직접 계산한 값을 테스트 해보는 것이 훨씬 더 명확하고 의미있는 것 같습니다! 앞으로는 이런 형태로 진행해야겠습니다!!
감사합니다!

stevenkim18 replied:
네네 항상 때에 따라 적절한 것을 사용해주세요 ㅎㅎ 이번 경우에는 위가 더 나아보였습니다!! ㅎㅎ

stevenkim18 reviewed (commented):
[](https://developer.apple.com/videos/play/wwdc2018/417/)마이노 스탭2도 고생많으셨습니다!

전반적으로 UML대로 너무 잘 구현하셨고 코드도 너무 깔끔했습니다.

앞으로의 마이노가 기대됩니다!!

### TDD

- 좋은 접근으로 테스트를 진행해보신 것 같아요!
- [https://developer.apple.com/videos/play/wwdc2018/417/](https://developer.apple.com/videos/play/wwdc2018/417/)
    - 위 영상에 2분 쯤에 보시면 테스트 단위가 나와요!
    - 큐, operator, formula는 unit test로 하신 거에요
    - 그리고 parser는 상대적으로 큰 단위를 테스트하신 것 같아요.
    - 항상 무엇이든 작은 단위로 시작해서 가능하면 endtoend 테스트 까지 가야 바람직하다고 생각해요.
    - 그러기 위해서는 설계도 중요한 것 같아요 클래스들을 잘 설계하면 테스트하기도 용이해 집니다 ㅎㅎ
- 저도 제대로 tdd를 해본적은 없지만 처음에는 시간이 많이 걸리는 것 같아 보이지만 결국 나중에는 애러 발생이 현지 줄어들어 오히려 더 시간이 단축되는 효과를 볼 수도 있다고 합니다. 테스트를 잘 구축해 놓으면 결국 자동화할 수 있는 것들이 많아서 오직 개발에만 집중할 수 있는 장점이 있다고 합니다(그런데 여전히 테스트와 테스트 하기 좋은 구조를 만들기는 늘 어렵습니다 ㅎㅎ)
- 0으로 나누는 로직과 같이 연결이 되어 있으면 저는 각각 테스트하고 최상위 메소드에서도 테스트 해주는 게 맞다고 봅니다.

**추가적인 내용**

![Untitled](https://user-images.githubusercontent.com/35272802/159252035-925c0469-8225-484a-a93f-2d170b635f46.png)

- 테스트 커버리지를 켜놓고 보니 디큐가 될 때 QueueError로 가지를 않네요.
- 이부분을 테스트 해봐주셨으면 합니다!
- 만약 다른 이유라면 이유를 찾아봐주세요!!!

### 메서드 분리

- 저도 처음 코드를 봤을 때는 복잡한(?) 느낌이 있네요 ㅎㅎ 고차함수를 많이 사용하셔서 그런 것 같아요.
- 맨 처음 고차함수를 사용하시지 않고 `componentsByOperators` 함수안에서만 모든 것을 처리해본 코드도 같이 첨부해주셨으면 마이노가 어떤 방향으로 리펙토링을 헀는지 더 감이 잘 왔을 것 같아요.
- 이번에는 조금 복잡하게 생각하시건 아닌가 싶어요. 만약 저라면 그냥 간단하고 보기 쉽게 구현했을 것 같아요!

Mino777 commented:
안녕하세요 스티븐! @stevenkim18 
피드백 반영해보았습니다!
좋은 리뷰 정말 감사합니다!!🥰🙇‍♂️

1. TDD `그리고 parser는 상대적으로 큰 단위를 테스트하신 것 같아요.`

해당 영상을 봐도 이해가 되지 않습니다😭  

해당 영상에서는 unit test를 지금처럼 앱의 개별 클래스와 메서드를 실행하는 코드레벨에 대한 테스트라고 지칭하는 것 같고 endtoend의 경우 UI테스트의 형태를 취하는 시스템 테스트로 지칭을 하는걸로 이해를 했습니다😭

그래서 큰 단위로 테스트를 진행했다고 말씀해주신 부분이 parse라는 메서드 하나로만 테스트를 진행했다고 말씀하신거로 이해를했는데 맞을까요?

parser의 경우 parse를 제외한 메서드들이 모두 private 메서드이기 때문에 암시적으로 테스트가 되게끔 진행을 했던 부분이었습니다.

해당 부분 한번만 더 설명 부탁드립니다😭🙇‍♂️

2. 테스트 커버리지를 켜놓고 보니 디큐가 될 때 QueueError로 가지를 않네요.

해당 부분은 생각해보니 result 메서드 안의 guard문에서 에러를 던지는 것이 아닌, 이미 dequeue를 진행하면서 notFoundElement 에러를 던졌기 때문에 첫 guard문에서 throw를 하기도 전에 바로 throw를 하기 때문에 테스트가 진행되지 않았던 것 같습니다. 제가 이해한게 스티븐의 의도와 맞을까요?

또한, dequeue의 결과값이 옵셔널인 부분도 에러처리로 인해 의미가 없어졌기 때문에 삭제하였습니다!

앞으로 테스트 커버리지로 커버가 안된 부분을 더블 체크하는 습관을 가져야될 것 같습니다. 감사합니다!!

3. 메서드 분리

```swift
static func parse(from input: String) -> Formula {
        var operands = CalculatorItemQueue<Double>()
        var operators = CalculatorItemQueue<Operator>()
        let operatorTypes = Operator
            .allCases
            .map { String($0.rawValue) }
        
        componentsByOperators(from: input)
            .compactMap { Double($0) }
            .forEach { operands.enqueue($0) }
        
        input.split(with: " ")
            .filter { operatorTypes.contains($0) }
            .compactMap { Operator(rawValue: Character($0)) }
            .forEach { operators.enqueue($0) }
                
        return Formula(operands: operands, operators: operators)
    }
    
    private static func componentsByOperators(from input: String) -> [String] {
        input.split(with: " ")
            .compactMap { Double($0) }
            .map { String($0) }
    }
```

위와 같이 진행을 하려고 하다가 너무 분리에 집착했던 것 같습니다. 앞으로는 분리를 하더라도, 가성비를 따지고 가독성을 살리면서 진행을 해보려고 노력해보겠습니다!
감사합니다!!

stevenkim18 commented:
> 테스트 커버리지를 켜놓고 보니 디큐가 될 때 QueueError로 가지를 않네요.
> 
- 네 저도 혹시 그런 것 같아서 확인 차 여쭤봤어요!!
- 저는 throw를 잘 사용하지는 않는데 여러번 쌓였을 때는 그냥 던진 곳에 끝나네요.
- 테스트 커버리지는 꼭 100을 하려고 하실 필요는 없습니다. 지금과 같은 애러를 잡거나 어디가 테스트가 안되었는지 확인할 떄 사용합니다.

> 메소드 분리
> 
- 또 이렇게 보니까 분리 해야 될 것 같기도 하구요... ㅋㅋ
- 근데 지금은 배우시는 시간이니까 다양하게 시도해보세요!! 커밋에 남고 리드미에 비교한 것 정리해 놓으시면 마이노의 자산이 될 거에요!!

stevenkim18 reviewed (approved):
테스트 관련해서는 구두로 말씀 드렸습니다!! 스탭3 하시면서 보완해 보시면 될 것 같아요. 고생하셨습니다!
머지할게요 ㅎㅎ
